### PR TITLE
Use a better heuristic for CPU kernel workgroup

### PR DESCRIPTION
**Pull request by glwagner**
The results over at https://github.com/glwagner/multithreaded-stencils suggest that we could be leaving as much as 2x on the table for CPU performance by choosing a non-optimal group size for multithreading.

This PR changes the group size on the CPU.

@hennyg888 could be good to runs some benchmarks to see if this improves our CPU / multithreading situation!

### CODE REVIEW

File: src/Utils/kernel_launching.jl [hunk 1]
```diff
@@ -6,10 +6,27 @@ using KernelAbstractions
...
 
-const MAX_THREADS_PER_BLOCK = 256
-
 flatten_reduced_dimensions(worksize, dims) = Tuple(i ∈ dims ? 1 : worksize[i] for i = 1:3)
 
+const MAX_GPU_THREADS_PER_BLOCK = 256
+
+function heuristic_workgroup(arch, Nx, Ny, Nz)
+    # One-dimensional column models:
+    Nx == 1 && Ny == 1 return (1, 1)
```
Comment (navidcy) [inline src/Utils/kernel_launching.jl:15]:
```suggestion
    Nx == 1 && Ny == 1 && return (1, 1)
```

??

**Review by navidcy — CHANGES_REQUESTED** [review]

File: src/Utils/kernel_launching.jl [hunk 2]
```diff
@@ -6,10 +6,27 @@ using KernelAbstractions
...
+    Nx == 1 && return (1, min(MAX_GPU_THREADS_PER_BLOCK, Ny))
+
+    # Two-dimensional x-z slice models:
+    Ny == 1 && return (1, min(MAX_GPU_THREADS_PER_BLOCK, Nx))
+
+    # Three-dimensional GPU models
+    arch isa GPU && return (Int(√MAX_GPU_THREADS_PER_BLOCK), Int(√MAX_GPU_THREADS_PER_BLOCK))
+
+    # Three-dimensional CPU models
+    return (Nx, Ny)
```
Comment (christophernhill) [inline src/Utils/kernel_launching.jl:27]: @glwagner I am trying to see why `return (Nx, Ny)` is the formula. 

Reading code here https://github.com/JuliaGPU/KernelAbstractions.jl/blob/545457ad3918b4d5aa687d6423f3e33909f368b6/src/KernelAbstractions.jl#LL348-L354 and below I got the impression that the workgroup size is the "granularity" that each thread will get. 

So if my granularity is Nx,Ny and I launch a kernel with worksize Nx,Ny,Nz I will spawn to at most Nz threads?  For some cases that could be 1 or some small number? 

To be fully general does it need to be Nx,Ny/NTHREADS or something more complicated involving Nz to ensure your workgroup size is < "data size"/NTHREADS.   Where "data size" is Nx*Ny*Nz - or it could be my reading of https://juliagpu.github.io/KernelAbstractions.jl/dev/ is confused!

For problems which mix 3d and 2d multithreading it might also be that some extra either dynamic or dispatch based variants are needed?

Comment (glwagner) [reply src/Utils/kernel_launching.jl:27]: > @glwagner I am trying to see why return (Nx, Ny) is the formula.

Only because it improved one benchmark, so it's worth discussing...

I think see what you're saying. Do you think we would want something like:

```julia
Nthreads = Base.Threads.threads()

y_residual_threads = max(1, Nthreads - Nz - 1)
By = floor(Int, Ny / y_residual_threads)

x_residual_threads = max(1, y_residual_threads - Ny - 1)
Bx = floor(Int, Nx / y_residual_threads)

return (Bx, By)
```

Thus for example if `Nz = 1`, then the problem is partitioned along `y`.

We could also add an explicit case for `Nz=1` and design an algorithm robust for `Nz > 1`.

@christophernhill your link may not have gone to the intended code block in `KernelAbstractions` --- can you send again?

@vchuravy maybe you have a recommendation?

Comment (christophernhill) [reply src/Utils/kernel_launching.jl:27]: @glwagner I think we need @vchuravy to clarify what WorkGroupSize means. 

The link is to where it is used, but it is hard to follow exactly how it then impacts the scheduling of CPU threads/GPU thread-blocks - so I may be misreading the code. My conclusion was that on a CPU `WorkGroupSize` ends up being the the size of the "blocks" the range is subdivided into for passing into different threads. I haven't tested that theory, so I could be wrong.

Comment (glwagner) [reply src/Utils/kernel_launching.jl:27]: The link goes to this docstring:

```
"""
    Kernel{Device, WorkgroupSize, NDRange, Func}
Kernel closure struct that is used to represent the device
kernel on the host. `WorkgroupSize` is the number of workitems
in a workgroup.
"""
```

Do you mean that the tuple is added to the type signature of `Kernel`?

Comment (glwagner) [reply src/Utils/kernel_launching.jl:27]: > @glwagner I think we need @vchuravy to clarify what WorkGroupSize means.
> 
> The link is to where it is used, but it is hard to follow exactly how it then impacts the scheduling of CPU threads/GPU thread-blocks - so I may be misreading the code. My conclusion was that on a CPU `WorkGroupSize` ends up being the the size of the "blocks" the range is subdivided into for passing into different threads. I haven't tested that theory, so I could be wrong.

Does that explain why we get speed up for 512^3 if we set (512, 512)? Basically this translates into splitting up the outer dimension?

Comment (christophernhill) [reply src/Utils/kernel_launching.jl:27]: Yeah - if my interpretation is correct then 512^3 would be fine, but 512x512x1 might be a problem. If we can clarify what `WorkGroupSize` means then coming up with new heuristic, if needed, should be straight forward.

Comment (christophernhill) [reply src/Utils/kernel_launching.jl:27]: I see @vchuravy said "yes" on Slack - so I think that means a 512x512x1 (or 2,3 etc...) problem would not scale without tweaking the heuristic.

Comment (glwagner) [reply src/Utils/kernel_launching.jl:27]: Here's my 🕵️ 

`workgroupsize` is used on the CPU here:

https://github.com/JuliaGPU/KernelAbstractions.jl/blob/51a59dc8ff2427b53db93a2e50efb4a972e370f5/src/cpu.jl#L106

to calculate `iterspace`. `iterspace` is used in `__run`:

https://github.com/JuliaGPU/KernelAbstractions.jl/blob/51a59dc8ff2427b53db93a2e50efb4a972e370f5/src/cpu.jl#L116-L137

possibly the logic in `__run` would limit us to one thread if `Nz=1`?

```
 N = length(iterspace)
    Nthreads = Threads.nthreads()
    if Nthreads == 1
        len, rem = N, 0
    else
        len, rem = divrem(N, Nthreads)
    end
    # not enough iterations for all the threads?
    if len == 0
        Nthreads = N
        len, rem = 1, 0
    end
```

Comment (vchuravy) [reply src/Utils/kernel_launching.jl:27]: Each task iterates over a number of blocks. The number of blocks is given by splitting the ndrange by the provided workgroupsize.

**Comment by glwagner** [comment]
@christophernhill checkout the new algorithm and let me know what you think

https://github.com/CliMA/Oceananigans.jl/blob/b15ac668a4f68b2c9841e69c1d7a1a725ff1c713/src/Utils/kernel_launching.jl#L27-L47

**Comment by glwagner** [comment]
@hennyg888 I think this is ready for benchmarking.

**Review by francispoulin — APPROVED** [review]
Changes look good and hope that the tests pass and benchmarks run efficiently

**Comment by francispoulin** [comment]
I guess here is a problem with `heuristic_workgroup`.

```
Time stepping NonhydrostaticModel: Time stepping NonhydrostaticModel: Error During Test at Time stepping NonhydrostaticModel: Time stepping NonhydrostaticModel: Error During Test at Error During TestError During Test at  at /var/lib/buildkite-agent/builds/tartarus-2/clima/oceananigans/test/test_distributed_models.jl:481
  | Got exception outside of a @test
  | MethodError: no method matching heuristic_workgroup(::MultiCPU{RegularRectilinearGrid{Float64, Periodic, Periodic, Periodic, OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}}}, Int64, Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Oceananigans.Distributed.RankConnectivity{Nothing, Nothing, Int64, Int64, Nothing, Nothing}, MPI.Comm}, ::Int64, ::Int64, ::Int64)
  | Closest candidates are:
  | heuristic_workgroup(::GPU, ::Any, ::Any, ::Any) at /var/lib/buildkite-agent/builds/tartarus-2/clima/oceananigans/src/Utils/kernel_launching.jl:12
  | heuristic_workgroup(::CPU, ::Any, ::Any, ::Any) at /var/lib/buildkite-agent/builds/tartarus-2/clima/oceananigans/src/Utils/kernel_launching.jl:26
 
```

**Comment by glwagner** [comment]
Might be good to see if this speeds things up before merging too.

**Comment by francispoulin** [comment]
> Might be good to see if this speeds things up before merging too.

I agree.  Now that the tests all pass, @henry888, would you mind giving this a try when you have a chance?

**Comment by hennyg888** [comment]
@glwagner the multithreading benchmarks for the 2d shallow water model does not seem to have changed. It's roughly the same as those shown in #1861.
<html>
<body>


size | threads | slowdown | efficiency | memory | allocs
-- | -- | -- | -- | -- | --
(8192, 512) | 1 | 1.0 | 1.0 | 1.0 | 1.0
(8192, 1024) | 2 | 1.98316 | 0.504246 | 12.1409 | 434.108
(8192, 2048) | 4 | 1.60839 | 0.621741 | 9.73477 | 331.135
(8192, 4096) | 8 | 1.52129 | 0.657337 | 10.2977 | 332.526
(8192, 8192) | 16 | 1.73345 | 0.576883 | 11.7069 | 340.188
(8192, 16384) | 32 | 2.26084 | 0.442313 | 17.1668 | 470.216


</body>
</html>

I'm also running another multithreading benchmark right now with the 3d nonhydrostatic model. Hopefully the new CPU workgroup grouping will show up better there.

**Comment by glwagner** [comment]
> @glwagner the multithreading benchmarks for the 2d shallow water model does not seem to have changed. It's roughly the same as those shown in #1861.
> 
> size	threads	slowdown	efficiency	memory	allocs
> (8192, 512)	1	1.0	1.0	1.0	1.0
> (8192, 1024)	2	1.98316	0.504246	12.1409	434.108
> (8192, 2048)	4	1.60839	0.621741	9.73477	331.135
> (8192, 4096)	8	1.52129	0.657337	10.2977	332.526
> (8192, 8192)	16	1.73345	0.576883	11.7069	340.188
> (8192, 16384)	32	2.26084	0.442313	17.1668	470.216
> I'm also running another multithreading benchmark right now with the 3d nonhydrostatic model. Hopefully the new CPU workgroup grouping will show up better there.

Thanks @hennyg888 !

I don't understand the non-monotonic behavior. What could explain this? It smells funny...

Does it make sense to run a strong scaling experiment as well to try to de-confuse the situation?

I wonder if we should actually precompute and store the thread-block layout in the `CPU` architecture type. That would allow us to more easily experiment with different layouts and also would allow users to tweak the layout if something wasn't working.

@christophernhill thoughts about that?

**Comment by hennyg888** [comment]
I just finished a strong scaling multithreaded benchmark with the nonhydrostatic model with grid size being 256 cubed.
I'll venture a wild guess and say that the weak scaling shallow water model's efficiency change is not monotonic due to smaller grid sizes not saturating the CPUs. Not sure if that's still a possible explanation when the grid size scales proportional to number of threads in weak scaling.

Here are the strong scaling nonhydrostatic results. I'm currently working on converting the script to do weak scaling to see if a small grid size with too many threads is what's causing very low efficiencies towards the end.
<html>
<body>


size | threads | slowdown | efficiency | memory | allocs
-- | -- | -- | -- | -- | --
256 | 1 | 1.0 | 1.0 | 1.0 | 1.0
256 | 2 | 0.992966 | 0.503542 | 4.14014 | 152.109
256 | 4 | 0.501089 | 0.498913 | 2.17724 | 50.2532
256 | 8 | 0.324366 | 0.385367 | 1.94899 | 29.191
256 | 16 | 0.244788 | 0.255323 | 2.12262 | 18.2106
256 | 32 | 0.263339 | 0.118668 | 2.87624 | 16.3167


</body>
</html>

**Comment by francispoulin** [comment]
The non-monotonic behaviour of the efficiency is very odd and certainly we shouldn't be getting that, if everything is set up correctly.

Strange that strong scaling looks more monotonic than the weak scaling.  I agree that on a coarser grid you don't necessary get as much speed up so those numbers probably aren't as significant as what happens on a finer grid.

**Comment by glwagner** [comment]
> I just finished a strong scaling multithreaded benchmark with the nonhydrostatic model with grid size being 256 cubed.
> I'll venture a wild guess and say that the weak scaling shallow water model's efficiency change is not monotonic due to smaller grid sizes not saturating the CPUs. Not sure if that's still a possible explanation when the grid size scales proportional to number of threads in weak scaling.
> 
> Here are the strong scaling nonhydrostatic results. I'm currently working on converting the script to do weak scaling to see if a small grid size with too many threads is what's causing very low efficiencies towards the end.
> 
> size	threads	slowdown	efficiency	memory	allocs
> 256	1	1.0	1.0	1.0	1.0
> 256	2	0.992966	0.503542	4.14014	152.109
> 256	4	0.501089	0.498913	2.17724	50.2532
> 256	8	0.324366	0.385367	1.94899	29.191
> 256	16	0.244788	0.255323	2.12262	18.2106
> 256	32	0.263339	0.118668	2.87624	16.3167

Excellent, nice results!

Same weird behavior --- the timing is identical from 1 to 2 threads; only at 4 threads do we get speed up (but its by a factor of 1/2 rather than 1/4). With 16 threads the model is 4x faster. Hmm... can we do better?

**Comment by hennyg888** [comment]
Might have something to do with how many threads are running on how many physical cores. I'll try to tune how the threaded processes are launched and specify one thread per core more clearly. Hyperthreading seems to be what enables 2 threads to run on one core. There should be options to disable it either through slurm or julia.

**Comment by glwagner** [comment]
Since there's no performance benefit, I'm closing this. We can revisit `heuristic_workgroup` in the future.
